src/stream/src/executor/over_window/general.rs

```diff
@@ -0,0 +1,925 @@
+
+        // Build final changes partition by partition.
+        for (part_key, diff) in diffs {
+            Self::ensure_partition_in_cache(this, &mut vars.partitions, &part_key).await?;
+            let mut partition = vars.partitions.get_mut(&part_key).unwrap();
+
+            // Build changes for current partition.
+            let part_changes = Self::build_changes_for_partition(
+                this,
+                DiffBTreeMap::new(partition.cache(), diff),
+            )?;
+
+            for (key, record) in part_changes {
+                // Build chunk and yield if needed.
+                if !key_change_updated_pks.contains(&key.pk) {
+                    if let Some(chunk) = chunk_builder.append_record(record.as_ref()) {
+                        yield chunk;
+                    }
+                } else {
+                    // For key-change updates, we should wait for both `Delete` and `Insert` changes
+                    // and merge them together.
+                    let pk = key.pk.clone();
+                    let record = record.clone();
+                    if let Some(existed) = key_change_update_buffer.remove(&key.pk) {
+                        match (existed, record) {
+                            (Record::Insert { new_row }, Record::Delete { old_row })
+                            | (Record::Delete { old_row }, Record::Insert { new_row }) => {
+                                // merge `Delete` and `Insert` into `Update`
+                                if let Some(chunk) =
+                                    chunk_builder.append_record(Record::Update { old_row, new_row })
+                                {
+                                    yield chunk;
+                                }
+                            }
+                            _ => panic!("other cases should not exist"),
+                        }
+                    } else {
+                        key_change_update_buffer.insert(pk, record);
+                    }
+                }
+
+                // Update state table and partition cache.
+                this.state_table.write_record(record.as_ref());
+                match record {
+                    Record::Insert { new_row } | Record::Update { new_row, .. } => {
+                        // If `Update`, the update is not a key-change update, so it's safe to just
+                        // replace the existing item in the cache.
+                        partition.insert(key, new_row);
+                    }
+                    Record::Delete { .. } => {
+                        partition.remove(&key);
+                    }
+                }
+            }
+        }
+
+        // Yield remaining changes to downstream.
+        if let Some(chunk) = chunk_builder.take() {
+            yield chunk;
+        }
+    }
+
+    fn build_changes_for_partition(
+        this: &ExecutorInner<S>,
+        part_with_diff: DiffBTreeMap<'_, StateKey, OwnedRow>,
```

Comment:
just pass the `snapshot` and `diff` here? maybe we do not need to expose `DiffBTreeMap` here